src/main/java/com/google/devtools/build/lib/analysis/BaseRuleClasses.java

```diff
@@ -345,10 +347,19 @@ public Metadata getMetadata() {
     return builder.add(attr("name", STRING).nonconfigurable("Rule name"));
   }
 
-  public static RuleClass.Builder execPropertiesAttribute(RuleClass.Builder builder)
+  /**
+   * Adds an {@code exec_properties} attribute of type {@code STRING_DICT} if the rule does not
+   * already have one.
+   */
+  public static void addOrOverrideExecPropertiesAttribute(RuleClass.Builder builder)
       throws ConversionException {
-    return builder.add(
-        attr(RuleClass.EXEC_PROPERTIES, STRING_DICT).defaultValue(ImmutableMap.of()));
+    Builder<Map<String, String>> attr =
+        attr(RuleClass.EXEC_PROPERTIES, STRING_DICT).defaultValue(ImmutableMap.of());
+    if (builder.contains(RuleClass.EXEC_PROPERTIES)) {
```

Comment:
What if the existing EXEC_PROPERTIES doesn't have the type string_dict? Should we throw an error here? Otherwise, it could be confusing to user.